core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -2160,6 +2234,43 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
         }
     }
 
+    // Visible for testing
+    long findLastOffsetAcknowledged() {
+        lock.readLock().lock();
+        long lastOffsetAcknowledged = -1;
+        try {
+            for (NavigableMap.Entry<Long, InFlightBatch> entry : cachedState.entrySet()) {
+                InFlightBatch inFlightBatch = entry.getValue();
+                if (inFlightBatch.offsetState() == null) {
+                    if (!isRecordStateAcknowledged(inFlightBatch.batchState())) {
+                        return lastOffsetAcknowledged;
+                    }
+                    // If initialReadGapOffset.gapStartOffset is less than or equal to the last offset of the batch
+                    // then we cannot identify the current inFlightBatch as acknowledged. All the offsets between
+                    // initialReadGapOffset.gapStartOffset and initialReadGapOffset.endOffset should always be present
+                    // in the cachedState
+                    if (initialReadGapOffset != null && inFlightBatch.lastOffset() >= initialReadGapOffset.gapStartOffset()) {
+                        return lastOffsetAcknowledged;
+                    }
+                    lastOffsetAcknowledged = inFlightBatch.lastOffset();
+                } else {
+                    for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+                        if (initialReadGapOffset != null && offsetState.getKey() >= initialReadGapOffset.gapStartOffset()) {
+                            return lastOffsetAcknowledged;
+                        }
+                        if (!isRecordStateAcknowledged(offsetState.getValue().state())) {
+                            return lastOffsetAcknowledged;
+                        }
+                        lastOffsetAcknowledged = offsetState.getKey();
+                    }
+                }
+            }
+        } finally {
+            lock.readLock().unlock();
+        }
+        return lastOffsetAcknowledged;
+    }
```

Comment:
Make the change inline with the earlier method to see the differences. We will update the ordering later.